### PR TITLE
RWatch Backend

### PR DESCRIPTION
**Pull request by 0xRishav**

### CODE REVIEW

File: controllers/Like.controller.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    const data = await user.execPopulate("likedVideos");
+    return res.status(200).json({ success: true, data: [...data.likedVideos] });
+  } catch (error) {
+    return res
+      .status(500)
+      .json({ success: false, message: "internal server error" });
+  }
+};
+
+module.exports.addToLikedVideos = async (req, res) => {
```
Comment (DanishOnGit) [inline controllers/Like.controller.js:16]: You can group these exports together,  `module.exports = {getLikedVideos,addToLikedVideos.....}` at the bottom of the file

File: controllers/Like.controller.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  const { userId, videoId } = req.params;
+  try {
+    const user = await userDb.findById(userId);
+    const likedVideos = user.likedVideos;
+    if (!likedVideos.some((video) => video == videoId)) {
+      await user.likedVideos.push(videoId);
```
Comment (DanishOnGit) [inline controllers/Like.controller.js]: You can remove the `await` for push statement

File: controllers/Playlist.controller.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
+const playlistDb = require("../models/playlist.model");
+const userDb = require("../models/user.model");
+
+module.exports.getAllPlaylists = async (req, res) => {
+  const { userId } = req.params;
+  try {
+    const user = await userDb.findById(userId);
+    console.log("usr Playlist", user.playlists);
```
Comment (DanishOnGit) [inline controllers/Playlist.controller.js]: You can clean the log statements if not needed

File: controllers/Playlist.controller.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    const { playlists } = await (
+      await userDb.findById(userId)
+    ).execPopulate({ path: "playlists", populate: { path: "videos" } });
```
Comment (DanishOnGit) [inline controllers/Playlist.controller.js:10]: Hey, I am unable to understand why we need to `await` statements. Can you please explain?

File: controllers/Playlist.controller.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    });
+  }
+};
+
+module.exports.addVideoToPlaylist = async (req, res) => {
+  const { playlistId, videoId } = req.params;
+  try {
+    const playlist = await playlistDb.findById(playlistId);
+    const userId = playlist.user;
+    console.log("PLAYLIST", playlist);
```
Comment (DanishOnGit) [inline controllers/Playlist.controller.js:84]: Here, we can use Alias for the same effect
`const {user : userId} = await playlistdb.findById(playlistId)`

File: controllers/Playlist.controller.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    if (!playlist.videos.includes(videoId)) {
+      playlist.videos.push(videoId);
+      playlist.save();
```
Comment (DanishOnGit) [inline controllers/Playlist.controller.js]: I think `playlist.save()` should be `await playlist.save()`

File: controllers/Playlist.controller.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    });
+  }
+};
+
+module.exports.removeVideoFromPlaylist = async (req, res) => {
+  const { playlistId, videoId } = req.params;
+  try {
+    const playlist = await playlistDb.findById(playlistId);
+    const userId = playlist.user;
+    await playlistDb.findByIdAndUpdate(playlistId, {
```
Comment (DanishOnGit) [inline controllers/Playlist.controller.js:113]: same Aliasing can be done here

File: controllers/User.controller.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
+const userDb = require("../models/user.model");
+
+module.exports.signUpUser = async (req, res) => {
+  const { name, email, password } = req.body;
+  let user = null;
+  try {
+    if (await userDb.findOne({ email: email })) {
+      return res.status(208).json({
+        success: false,
+        message: "This email is already registered",
```
Comment (DanishOnGit) [inline controllers/User.controller.js]: Have you considered using the 409 status code here ?

File: controllers/User.controller.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
+const userDb = require("../models/user.model");
+
+module.exports.signUpUser = async (req, res) => {
+  const { name, email, password } = req.body;
+  let user = null;
+  try {
+    if (await userDb.findOne({ email: email })) {
```
Comment (DanishOnGit) [inline controllers/User.controller.js]: Here we can refactor:
```
const user = await userDb.findOne({email:email})
try{
if(user){....}
}
```

File: controllers/history.controller.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      .status(500)
+      .json({ success: false, message: "internal server error" });
+  }
+};
+
+module.exports.addToHistory = async (req, res) => {
+  const { userId, videoId } = req.params;
+  try {
+    const user = await userDb.findById(userId);
+    const history = user.history;
```
Comment (DanishOnGit) [inline controllers/history.controller.js:20]: here `const {history} =await userdb.findById(userId)` should be enough

File: controllers/history.controller.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    if (!history.some((video) => video == videoId)) {
+      await user.history.push(videoId);
```
Comment (DanishOnGit) [inline controllers/history.controller.js]: `await` for push can be removed.  Also, you have `history` available above.  So, `history.push(videoId)` can be written

File: controllers/history.controller.js [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    let user = await userDb.findById(userId);
+    if (user.history.includes(videoId)) {
+      await user.update({ $pull: { history: videoId } });
+    } else {
+      return res
+        .status(400)
+        .json({ success: false, message: "video is not liked" });
+    }
+
+    const newUser = await userDb.findById(userId);
```
Comment (DanishOnGit) [inline controllers/history.controller.js:51]: You can try doing this: `const newUser = await userDb.findById(userId).populate("history")`

File: controllers/history.controller.js [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    return res.status(200).json({ success: true, data: [...data.history] });
+  } catch (error) {
+    return res
+      .status(500)
+      .json({ success: false, message: "internal server error" });
+  }
+};
+
+module.exports.resetHistory = async (req, res) => {
+  console.log("inside history controller")
```
Comment (DanishOnGit) [inline controllers/history.controller.js]: It's good to remove all logs which are not needed

File: controllers/history.controller.js [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+  const { userId } = req.params;
+  try {
+    let user = await userDb.findById(userId);
+    console.log("found user", user.email);
+    user.history = [];
+    console.log("updated history", user.history);
```
Comment (DanishOnGit) [inline controllers/history.controller.js]: Here too, the logs can be cleaned

File: models/history.model.js [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+const historyVideoSchema = new Schema(
+  {
+    user: {
+      type: Schema.Types.ObjectId,
+      ref: "User",
+    },
+    historyVideo: {
+      video: mongoose.Types.ObjectId,
+      ref: "Video",
+    },
```
Comment (DanishOnGit) [inline models/history.model.js]: Shouldn't this be : video:`Schema.Types.ObjectId` ?

**Review by DanishOnGit — COMMENTED** [review]
Hey Rishav, I have suggested some changes about refactoring,status code, push methods .

File: config/mongoose.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+const uri = process.env.URI_KEY;
+
+mongoose
+  .connect(uri, {
+    useNewUrlParser: true,
+    useUnifiedTopology: true,
+  })
+  .then(() => console.log("mongoose connection successful"))
+  .catch((err) => console.log("mongoose connection failed", err));
+
```
Comment (neharajesh) [inline config/mongoose.js:13]: Using try/catch and async/await is a better practice!

File: controllers/history.controller.js [hunk 17]
```diff
@@ -0,0 +1,81 @@
+const userDb = require("../models/user.model");
+
+module.exports.getHistory = async (req, res) => {
+  const { userId } = req.params;
+  try {
+    const user = await userDb.findById(userId);
+    const data = await user.execPopulate("history");
+    return res.status(200).json({ success: true, data: [...data.history] });
```
Comment (neharajesh) [inline controllers/history.controller.js:8]: Keeping your json responses consistent would help when you're fetching/working with the data! Most of the previous responses have a message, adding one here would be good!

File: .gitignore [hunk 18]
```diff
@@ -0,0 +1,23 @@
+.env
+requests.http
+# dependencies
+/node_modules
```
Comment (neharajesh) [inline .gitignore:4]: Node modules haven't been removed, do check!

**Review by neharajesh — COMMENTED** [review]
Hey Rishav
I've suggested some changes, hope this helps!

**Pull request merged**